_posts/2020-10-28-push-it-to-the-limit.md

```diff
@@ -0,0 +1,58 @@
+
+Back of the envelope math said this shouldn't be too much of a problem, A highly optimistic 200 users pulling down 350 bytes of data once a minute was only 100MB a day, leaving us a 2x safety net versus our 10GB monthly limit. Trusting that we did the simplest implementation possible and just let users poll the data each minute and went on with working on other things, and we were correct for several years.
+
+## So We Were Wrong, Now What?
+
+So it became obvious that we'd missed something, and it wasn't until I hovered over the bandwidth tooltip in the site console that I realized where we'd gone awry.
+
+> Downloads account for all bytes downloaded from your database, including connection protocol and SSL encryption overhead
+
+But how much overhead could those really entail? I didn't have an easy way to measure it from our metrics nor did we really know how many times users were requesting data (or so I thought at the time). Instead I had a potential solution that was probably faster to implement than it would be to measure out what our per communication overhead actually was so I went with that approach.
+
+Everyone who's worked with web development is familiar with minified files and why we do that, well I figured it was time to minify the footprint of our data json and I could just apply mappings locally to restore human readable values.
+
+![original moon data format](/assets/moon_data_v1.PNG) ![minified moon data format](/assets/moon_data_v2.PNG)
+
+These changes took our data from 350 bytes down to 200, even if the communication overhead was doubling our data usage we should still see a 10-15% reduction, keeping us under our limit for probably another year of user growth easily. On May 3rd I flipped the switch and patiently waited for the metrics to roll in, and in one week I planned on disabling the old data endpoint entirely.
+
+![bandwidth usage chart over time](/assets/BlogBandwidthNoChange.PNG)
+
+And... no measurable change occurred over the next several days. Perhaps most of our users weren't on the new codebase? On the 9th I killed the old data endpoint, still nothing. While I still didn't know exactly how bad the communication overhead was I'd proven that no change to our data format was going to resolve the issue, we needed to be smarter about requesting data.
+
+## Well That Was Easy
+
+Perhaps not surprisingly, the correct solution really wasn't that difficult to implement. We had decent estimates on when the next event should occur, why don't we trust those estimates a little more? They may be a few minutes off over a 3 hour window, but they're never going to be an hour off. We changed our code to instead of blindly polling once a minute, to poll at the greater of either 80% of the time until the next event, or 1 minute. We'd keep the same accuracy, but reduce the number of requests dramatically. On may 16th I made this change and pushed it out to the users.
```

Comment:
> On may 16th

There are a few places proper nouns aren't capitalized. Consider making another editing pass for that.